R/calculate_study_day.R

```diff
@@ -0,0 +1,133 @@
+#' `calculate_study_day` performs study day calculation
+#' @description
+#' This function takes the an input data frame and a reference data frame (which
+#' is DM domain in most cases), and calculate the study day from reference date
+#' and target date. In case of unexpected conditions like reference date is not
+#' unique for each patient, or reference and input dates are not actual dates,
+#' NA will be returned for those records.
+#'
+#' @param sdtm_in Input data frame that contains the target date.
+#' @param tgdt Target date from `sdtm_in` that will be used to calculate the study
+#' day.
+#' @param study_day_var New study day variable name in the output. For
+#' example, AESTDY for AE domain and CMSTDY for CM domain.
+#' @param dm_domain Reference date frame that contains the reference date.
+#' @param refdt Reference date from `dm_domain` that will be used as reference to
+#' calculate the study day.
+#' @param merge_key Character to represents the merging key between `sdtm_in` and
+#' `dm_domain`.
+#'
+#' @return Data frame that takes all columns from `sdtm_in` and a new variable
+#' to represent the calculated study day.
+#'
+#' @export
+#'
+#' @examples
+#' ae <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   AESTDTC = c("2012-01-01", "2012-04-14", "2012-04-14")
+#' )
+#' dm <- data.frame(
+#'   USUBJID = c("study123-123", "study123-124", "study123-125"),
+#'   RFSTDTC = c("2012-02-01", "2012-04-14", NA)
+#' )
+#' ae$AESTDTC <- as.Date(ae$AESTDTC)
+#' dm$RFSTDTC <- as.Date(dm$RFSTDTC)
+#' calculate_study_day(ae, "AESTDTC", "AESTDY", "dm", "RFSTDTC")
+#'
+calculate_study_day <- function(sdtm_in,
+                                tgdt,
+                                study_day_var,
+                                dm_domain = "DM",
+                                refdt = "RFSTDTC",
+                                merge_key = "USUBJID") {
+  assertthat::assert_that(is.data.frame(sdtm_in))
+  dm_domain <- eval(parse(text = dm_domain))
```

Comment:
Hi @ramiromagno @rammprasad @ShiyuC If I default dm_domain to "DM", I will need to evaluate "DM" to the object in the environment where the function was made. This is expected however unit testing will fail as the function is wrapped in test_that() or expect_warning(). How do I fix this?

Reply:
Hi Rosemary,

I think we should really not have a default value for the `dm_domain` parameter. The data frame to be passed should be passed during the call, i.e. we should not be looking up in the environment some object referenced by name, be it a bare name or provided as a string.

```suggestion
calculate_study_day <- function(
                                sdtm_in,
                                tgdt,
                                study_day_var,
                                dm_domain,
                                refdt = "RFSTDTC",
                                merge_key = "USUBJID") {
  assertthat::assert_that(is.data.frame(sdtm_in))
  # dm_domain <- eval(parse(text = dm_domain))
  assertthat::assert_that(is.data.frame(dm_domain))
```

Then change the unit tests such that the function calls take directly the `dm` data frame, e.g.:

```R
  expect_warning(
    calculate_study_day(ae, "AESTDTC", "AENDY", dm_domain = dm, "RFSTDTC"),
    "Target date and the returned study day doesn't match."
  )
```

Reply:
Hi @ramiromagno I was told to default this arg to DM domain from previous discussion, that's why I was defaulting to `DM` instead of `"DM"` in the beginning, but not sure which is preferred.

@rammprasad @edgar-manukyan @ShiyuC Can you guys chime in please?

Reply:
We discussed defaulting it to the DM domain in the meeting. But I am ok with not defaulting it as DM or dm. It could get into the intricacies of uppercase `DM` or lower case `dm`

Reply:
Hi @ramiromagno and @rammprasad I can remove the default but I really want to know **if we have the default, how do I fix the unit testing failure? I think this will help me have a better understanding of R environments**, so please advise? Meanwhile I can remove the default

Reply:
Hi @yli110-stat697 and @rammprasad : I think we had agreed early on to not rely on having variables being looked up in  implicit environments. So just to make clear, my recommendation is not specific about the DM domain, but more generally about any function argument. What I am suggesting as alternative is to have the data frame corresponding to the DM domain data set passed by value. If understand correctly, calling:

```R
calculate_study_day(ae, "AESTDTC", "AENDY", dm_domain = dm, "RFSTDTC")
```

does not generate any issues with the unit testing, right? Because the `dm` data frame is defined before.

Reply:
BTW if we don't default dm_domain, do we still need to default `refdt`? @rammprasad @ramiromagno

Reply:
> does not generate any issues with the unit testing, right? Because the dm data frame is defined before.

@ramiromagno if we don't default dm_domain, or default to object `DM` or `dm`, yes we don't have issues. If we default to `"DM"` (the string), calling the functions still works, but the unit testing failed. Because I used 
`dm_domain <- eval(parse(text = dm_domain))` to evaluate "DM" string to DM object. 

My question is **if we default to "DM" string, is there better way to do this so that the unit testing can pass**?

Reply:
In this specific case, we can make the unit test pass with this code:

```R
dm_domain <- eval(parse(text = dm_domain), envir = parent.frame(n = 1))
```

The reason why you need the `parent.frame(n = 1)` is so that `eval()` goes looking in the parent environment of `expect_warning()`, and not in the environment of `expect_warning()` itself, where the `dm` object is not defined.

But this is very much not recommended. If you go down this path, then your code will be context dependent, i.e. it will depend in which context (i.e. environment) `calculate_study_day()` is called in. This defies the whole point of having a function which is to encapsulate the computation with a clear set of inputs while being isolated from the rest of your program context.

This issue illustrates the exact reason why I am recommending against looking up variables based on their names.